### PR TITLE
Wamp session interface

### PR DESCRIPTION
**Pull request by om26er**
First, sorry for the missing comments. Will add them once the interface is looking good.

Playground.java is what shows how the interaction is supposed to be. I have added a new `wamp` package and added `types` and `interfaces` as its sub-packages; trying to keep inline with ABPy.

### CODE REVIEW

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/Playground.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import io.crossbar.autobahn.wamp.types.CallResult;
+import io.crossbar.autobahn.wamp.types.PublicationResult;
+import io.crossbar.autobahn.wamp.types.Registration;
+import io.crossbar.autobahn.wamp.types.Subscription;
+
+public class Playground {
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/Playground.java]: Yes, Java 8 lambdas now and seems they are pretty solid.

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import io.crossbar.autobahn.wamp.types.ISubscribeHandler;
+import io.crossbar.autobahn.wamp.types.PublicationResult;
+import io.crossbar.autobahn.wamp.types.PublishOptions;
+import io.crossbar.autobahn.wamp.types.RegisterOptions;
+import io.crossbar.autobahn.wamp.types.Registration;
+import io.crossbar.autobahn.wamp.types.SubscribeOptions;
+import io.crossbar.autobahn.wamp.types.Subscription;
+
+public interface ISession {
+    int STATE_CONNECTED = 1;
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java]: done

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    int STATE_DISCONNECTED = 3;
+    int STATE_DISCONNECTED_RESUMABLE = 4;
+    int STATE_HELLO_SENT = 5;
+    int STATE_AUTHENTICATE_SENT = 6;
+    int STATE_JOINED = 7;
+    int STATE_READY = 8;
+    int STATE_GOODBYE_SENT = 9;
+    int STATE_SENT = 10;
+
+    CompletableFuture<Subscription> subscribe(ISubscribeHandler handler, String topic, SubscribeOptions options);
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java]: done.

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    CompletableFuture<PublicationResult> publish(String topic,
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java]: done

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                                                 List<Object> args,
+                                                 Map<String, Object> kwargs,
+                                                 PublishOptions options);
+    CompletableFuture<Registration> register(IEndpoint endpoint, String procedure, RegisterOptions options);
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java]: done.

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package io.crossbar.autobahn.wamp.interfaces;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import io.crossbar.autobahn.wamp.types.CallOptions;
+import io.crossbar.autobahn.wamp.types.CallResult;
+import io.crossbar.autobahn.wamp.types.IEndpoint;
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/interfaces/ISession.java]: yeah, this part is a bit unclear, will chat about that.

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/types/IEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package io.crossbar.autobahn.wamp.types;
+
+import java.util.List;
+import java.util.Map;
+
+@FunctionalInterface
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/types/IEndpoint.java]: This basically allows us to pass in anonymous methods as parameters; new in Java8. Any method having this as a parameter will ensure that the callback method that we pass in have exactly the same signature.

In this case it helps our API to only except a method as callback that has the signature
`List<Object> args, Map<String, Object> kwargs`

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/types/IEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+public interface IEndpoint<R> {
+    // Tentative name, to be changed.
+    R get(List<Object> args, Map<String, Object> kwargs);
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/types/IEndpoint.java]: the `get` here is basically the callback i.e. calling `get` will actually call the callback that we passed in.

We definitely need to change its from `get` to something sane.

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/types/CallResult.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package io.crossbar.autobahn.wamp.types;
+
+import java.util.List;
+import java.util.Map;
+
+public class CallResult {
+    public List<Object> results;
+    public Map<String, Object> kwresults;
+
```
Comment (om26er) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/types/CallResult.java]: Sorry, I am not able to understand, can you re-elaborate on what you mean here.

**Comment by om26er** [comment]
Thanks for your review, I have made the suggested changes and also replied inline.

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/types/CallResult.java (hunk 9 above)
Comment (oberstet) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/types/CallResult.java]: copy constructor would take another CallResult and create a new CallResult with the values deep copied (not merely copying the references) - it's a minor point. we can add that later if needed.

**Pull request merged**